File: hanjo/개인용/programmers/구현/L2_프린터/solution.py
from collections import deque

def solution(priorities, location):
    # [인덱스, 우선순위]로 그룹핑 해준 후 큐로 변환
    p_list = deque([[i,p] for i, p in enumerate(priorities)])
    # 우선순위도 높은 순 정렬 후 큐로 변환
    priorities = deque(sorted(priorities, reverse=True))
    cnt = 0
    while 1:
        # 1. 맨 앞에 있는 프린터를 꺼낸다
        temp = p_list.popleft()
        # 2. 우선순위 큐의 가장 큰 것과 순위 비교.
        if priorities[0] != temp[1]:
            # 3. 해당 프린터보다 우선순위가 더 큰게 있다면 뒤로감
            p_list.append(temp)
        else:
            # 4. 해당 프린터의 우선순위가 가장 크면 우선순위 pop.
            cnt+=1
            priorities.popleft()
            # 5. 만약 그게 내가 구하고싶은 프린터라면?
            if temp[0] == location:
                break
    return cnt


priorities = [2, 1, 3, 2]
location = 2
print(solution(priorities, location))

File: hanjo/개인용/programmers/구현/L2_프린터/Solution.java
package 구현.L2_프린터;

import java.util.*;

public class Solution {
	public static int solution(int[] priorities, int location) {
		
		// 배열 -> arraylist
		List<int[]> list = new ArrayList<>();
		// 우선순위를 내림차순으로 정렬
		PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
		// 리스트와 큐 초기화
		for(int i=0; i<priorities.length; i++) {
			list.add(new int[]{i, priorities[i]});
			pq.add(priorities[i]);
		}
		
		int count = 0;
		while(pq.size() > 0) {
			count ++;
			int highestPriority = pq.poll();
			while(true) {
				int[] temp = list.remove(0);
				int tempLocation = temp[0];
				int tempPriority = temp[1];
				
				if(tempPriority == highestPriority) {
					if(tempLocation == location) {
						return count;
					}
					break;
				}
				else {
					list.add(new int[]{tempLocation, tempPriority});
				}
			}
			
		}
		return count;
    }

	public static void main(String[] args) {
		int[] priorities = {2, 1, 3, 2};
		int location = 2;
		System.out.println(solution(priorities, location));
	}
}